lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -103,6 +103,18 @@ def replace_metadata(resource, metadata, recurse = false)
         resource[:checksum] = metadata.checksum_type
       end
     end
+
+    # The static file content server doesn't know how to expand
+    # mountpoints, so we need to do that ourselves from the actual
+    # system math of the source file. This does that, while preserving
+    # any user-specified server or port.
+    source_path = Pathname.new(metadata.full_path)
+    environment_path = Pathname.new(environment_path)
+    path = source_path.relative_path_from(environment_path).to_s
+    source_as_uri = URI.parse(metadata.source)
```

Comment:
I think we need to:

```ruby
      uri = URI.parse(URI.escape(metadata.source))
```

as I believe `metadata.source` is not URL encoded, and is why we do https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/file/source.rb#L73. For example if you have in your manifest:

```puppet
file { '/tmp/foo bar':
  ensure => file,
  source => 'puppet:///modules/mymodule/foo bar'
}
```

Then `metadata.source` will contain a space, and trying to parse it will fail:

```
(byebug) metadata.source
"puppet:///modules/ntp/foo bar"
(byebug) URI.parse(metadata.source)
*** URI::InvalidURIError Exception: bad URI(is not URI?): puppet:///modules/ntp/foo bar
```

Secondary issue, `URI.escape` is deprecated, and we should probably be using `CGI.escape` instead. http://stackoverflow.com/questions/2824126/whats-the-difference-between-uri-escape-and-cgi-escape That's probably worthy of a separate non-direct puppet related ticket, as we do URI.escape in several places.